### PR TITLE
Implement event-specific pooling for SyntheticEvent

### PR DESCRIPTION
**Pull request by aweary**
WIP. I think this is mostly done, still looking at pooling in `TopLevelCallbackBookKeeping`

This PR removes `PooledClass` from the synthetic event system in favor of a simpler, event-specific pooling system. 

* It maintains the same pooling API (`getPooled`, `release`) to minimize churn in the event system.
* Instead of `PooledClass.addPoolingTo` it now uses a helper `addEventPoolingTo` defined in the `SyntheticEvent` module
* Removes pooling all-together from `FallbackCompositionState`. It was only being used in a single event plugin, and the way it's structured makes it so there's only ever a single instance anyway. This means we don't need any sort of pooling. Instead, the composition state is now tracked in a single object which can be initialized and reset.

### CODE REVIEW

**Comment by gaearon** [comment]
Nice! Can you also get it out of `ReactDOMEventListener`? Then we can move it to the Stack folder.

**Comment by aweary** [comment]
Yupp! That's the one last thing I have to do, will remove the WIP tag and ping when I get that done 😄

**Comment by jquense** [comment]
neat! What is the benefit of splitting these two out if I might ask.

**Comment by aweary** [comment]
@jquense there's some context here: https://github.com/facebook/react/issues/9325 The goal is to get rid of `PooledClass` all together in Fiber, in favor of a more focused approach. For example, we can now consider implementing a much more targeted pooling strategies for events (one instance per event type).

It does lead to a bit of duplication, but I also think it makes it clearer what's going on.

**Comment by aweary** [comment]
We also end up with a _slightly_ smaller build as well, since `PooledClass` already contained a bunch of duplication with the various poolers it exports.

**Comment by gaearon** [comment]
>slightly larger react bundle

Why?

**Comment by aweary** [comment]
I'm not sure, I'll look into it though.

**Comment by gaearon** [comment]
I wouldn't expect changes in your branch to affect the size. It's just that we don't record sizes on every commit so they fluctuate. If you want an accurate measurement of your commit specifically, you can roll back to merge base, record sizes there, then apply your changes and record them again.

**Comment by aweary** [comment]
Ah yeah, you're right @gaearon, the current size report is just out of date. With this change the `react` size remains the same at `2759` and `react-dom` goes down from `40030` to `39828`.

**Comment by gaearon** [comment]
Sounds good. Is this still WIP?

**Comment by aweary** [comment]
I think it's ready for review!

File: src/renderers/dom/shared/ReactDOMEventListener.js [hunk 1]
```diff
@@ -22,6 +21,9 @@ var getEventTarget = require('getEventTarget');
 
 var {HostRoot} = ReactTypeOfWork;
 
+var CALLBACK_BOOKKEEPING_POOL_SIZE = 10;
```
Comment (nhunzaker) [inline src/renderers/dom/shared/ReactDOMEventListener.js]: Why 10?

Comment (gaearon) [inline src/renderers/dom/shared/ReactDOMEventListener.js]: It was default in the old code.

**Comment by gaearon** [comment]
I'm good with this but please verify that pooling works as expected by manual inspection. E.g. make a page listening to mousemove and verify the values are up to date but instances are reused. And that persist() still works.

**Comment by nhunzaker** [comment]
I added some event pooling test fixtures:
http://pooled-event-fixtures.surge.sh/event-pooling

Curious what the best manual test of this is, but I found these two tests helpful when throwing breakpoints in the source.

**Comment by gaearon** [comment]
Could you please fix conflicts?

**Comment by aweary** [comment]
@gaearon conflicts resolved 🚀

**Comment by gaearon** [comment]
Hmmm. CI still failing.

**Comment by gaearon** [comment]
This looks good to me. Previously, only Prettier failed so I’m merging it without waiting for CI. (I fixed Prettier).

**Comment by aweary** [comment]
Thanks for fixing that @gaearon!

**Comment by gaearon** [comment]
Thanks for doing this ❤️

**Pull request merged**